networkx/algorithms/traversal/breadth_first_search.py

```diff
@@ -370,6 +371,53 @@ def bfs_successors(G, source, depth_limit=None, sort_neighbors=None):
     yield (parent, children)
 
 
+def bfs_layers(G, sources):
+    """Returns an iterator of all the layers in breadth-first search traversal.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        A graph over which to find the layers using breadth first search.
+
+    sources : list of nodes in `G`
+        Specify starting nodes for single source or multiple sources breadth-first search
+
+    Yields
+    -------
+    layer: list of nodes
+        Yields list of nodes at the same distance from sources
+
+    Examples
+    --------
+    >>> G = nx.path_graph(5)
+    >>> dict(enumerate(nx.bfs_layers(G, [0, 4])))
+    {0: [0, 4], 1: [1, 3], 2: [2]}
+    >>> H = nx.Graph()
+    >>> H.add_edges_from([(0, 1), (0, 2), (1, 3), (1, 4), (2, 5), (2, 6)])
+    >>> dict(enumerate(nx.bfs_layers(H, [1])))
+    {0: [1], 1: [0, 3, 4], 2: [2], 3: [5, 6]}
+    >>> dict(enumerate(nx.bfs_layers(H, [1, 6])))
+    {0: [1, 6], 1: [0, 3, 4, 2], 2: [5]}
+    """
+    for source in list(sources):
+        if not G.has_node(source):
```

Comment:
A good way to check if a node is in a graph is "node in G".  (this doesn't work with edges... only nodes).
A list (or dict or other unhashable) object cannot be a node. And a tuple can be but could also be a container of nodes. So a good type-agnostic way of checking whether the input is a list or a single node is to check if the thing is in G.  So, we check `if sources in G:` and if it is, we consider it to be a single node. We want an iterable of nodes so we wrap it in a list.  Note that this works pretty well for strings too. If input "abcd" is a node in G, it is wrapped in a list -- which seems reasonable. If it is not a node, then it is treated as a container of nodes `['a', 'b', 'c', 'd']`.  That is also reasonable.

After we set up the sources object, we iterate over the sources and check if each is in G, raising an exception if they are.  Finally, we use `sources` to fill up the current_layer, and visited. 

Idea:  This doesn't allow us to handle an iterator.  It would be exhausted after the first loop and not be able to fill the current_layer.  So, let's make this work for either a node, iterable or iterator.  We'll fill the current_layer before we iterate over it to check that the nodes are in G.   What do you think:
```
    # check for single node
    if sources in G:
        sources = [sources]

    current_layer = list(sources)
    visited = set(current_layer)

    for source in current_layer:
        if source not in G:
            raise nx.NetworkXError(...)

```